wandb/sdk/internal/internal_api.py

```diff
@@ -1040,13 +1037,74 @@ def create_run_queue(
         ]
         return result
 
+    @normalize_exceptions
+    def push_to_run_queue_by_name(
+        self, entity: str, project: str, queue_name: str, run_spec: str
+    ) -> Optional[Dict[str, Any]]:
+        """
+        Queryless mutation, should be used before legacy fallback method
+        """
+
+        mutation = gql(
+            """
+        mutation pushToRunQueueByName(
+            $entityName: String!,
+            $projectName: String!,
+            $queueName: String!,
+            $runSpec: JSONString!,
+        ) {
+            pushToRunQueueByName(
+                input: {
+                    entityName: $entityName,
+                    projectName: $projectName,
+                    queueName: $queueName,
+                    runSpec: $runSpec
+                }
+            ) {
+                runQueueItemId
+            }
+        }
+        """
+        )
+        variables = {
+            "entityName": entity,
+            "projectName": project,
+            "queueName": queue_name,
+            "runSpec": run_spec,
+        }
+        result: Optional[Dict[str, Any]] = self.gql(mutation, variables).get(
+            "pushToRunQueueByName"
+        )
+
+        return result
+
     @normalize_exceptions
     def push_to_run_queue(
         self, queue_name: str, launch_spec: Dict[str, str]
     ) -> Optional[Dict[str, Any]]:
-        # TODO(kdg): add pushToRunQueueByName to avoid this extra query
         entity = launch_spec["entity"]
         project = launch_spec["project"]
+        run_spec = json.dumps(launch_spec)
+
+        try:
+            push_result = self.push_to_run_queue_by_name(
+                entity, project, queue_name, run_spec
+            )
+        except CommError as e:
+            # explicitly handle servers with no pushToRunQueueByName mutation
+            if (
+                'Cannot query field "pushToRunQueueByName" on type "Mutation"'
+                in e.message
+            ):
```

Comment:
Do you think there is a better way to handle this case. I don't think it is a good idea to rely on string matching in the error message, it could easily change in the backend and we have no guards against it, unless there are relevant tests in core as well.
Thus far we either relied on introspection (but it means one extra query and it seems that introspection is not encouraged in production, seems to be some security concerns) or similar to what you did here, basically tried the query and if it failed we tried another one.

Reply:
Navigating this is pretty tricky, the way that Kyle nudged me toward in the past was a simple, but crude, try/catch like this. I agree it's definitely not elegant, but the error message is pretty standard; I can't imagine we change the backend language around GQL errors very frequently (if ever)...

Reply:
`upsert_sweep` has some similar patterns for trying a cascading set of mutations. There they specifically check:
`usageError` and `Exception` where Exception is used to check a graphql schema exception. Note that they specifically check to not retry 4xx errors, which we would also want to do here potentially. Maybe @raubitsj can provide more info on error handling for falling back to legacy mutations?